### DISCUSSION TITLE
Implementing Code Completion for a database - how should the code behave?

### DISCUSSION THREAD

**Discussion by wernerdaehn** [dbeaver/dbeaver#11162] [DBeaver Development]
I had filed a few issues around code completion for the Hana database and debugged the code. I can see how it works but at a few places I don't know how DBeaver is supposed to work. Also looked at the Oracle implementation for reference but some questions are open still. If you can help me to understand what DBeaver should do for all databases, I can bring SAP Hana up to the standard. And if something is not implemented in DBeaver yet, I must no longer search for the non-existing code.

* Caching of the suggestion list: In my point of view the cache should be updated at certain cases but is not
  * Application restart: It seems the list is cached across application restarts. Is that a good idea? How to change the default? Where is it cached?
  * A drop/create table/view/synonym should modify the cache. Does it for other databases?
  * How can a user force the suggestion list to be invalidated and reloaded? Found no traces of that.
  * When the user does refresh the database navigator, shouldn't that update the impacted parts of the suggestion list?
* Object types: The suggestion list is called with an object type but what should it return?
  * Per my finding the suggestion list for TABLE should return all objects a user can use in select-from clauses. So tables, views, synonyms. Correct?
  * That raises the question on when the type VIEW is provided. The only use case I can see is when a user types "drop view ...". Then only views make sense. Is that the desired logic? In Oracle I have not seen that.

Thanks in advance!

**Comment by jhdba** [comment]
I also really miss an easy way of refreshing the suggestion list. The only way I know is to refresh the connection and then reattach it to the script, but it is really anoying.

**Comment by wernerdaehn** [comment]
I am still trying to find my way around that. When I pulled the most recent version yesterday, I believe the code works differently now. The HANAStructureAssistant is no longer called at all but instead the entire schema from the database navigator is used.
Actually, I believe that is the better approach anyhow. It means a refresh of the schema does refresh code completion, once a drop table SQL changes the database navigator it changes the autocomplete list as well.
Frankly it solves the entire problem space in fact. 
I will play around with that some more, check if all Hana objects suited for auto completion are present and maybe make code changes.

@serge-rider Can you please confirm that my findings are correct and I am heading into the proper direction? Still trying to find my way...
